tls: Upgrade to Mbed TLS v2.25.0

Patater opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

Upgrade to Mbed TLS v2.25.0

Please refer to the Mbed TLS v2.25.0 release notes at https://github.com/ARMmbed/mbedtls/releases/tag/v2.25.0 for more information about what comes with this release.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
None

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
None

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->
None

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [X] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [X] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

Patater committed:
tls: Upgrade to Mbed TLS v2.25.0

LDong-Arm reviewed (commented):
Should be good as everything was imported using the script. Waiting for manual testing.

LDong-Arm commented on platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_MBED_PSA_SRV/mbedtls/psa_crypto.c:
```diff
@@ -5946,8 +6491,6 @@ static psa_status_t psa_generate_key_internal(
-        if( curve_info->bit_size != bits )
-            return( PSA_ERROR_INVALID_ARGUMENT );
         mbedtls_ecp_keypair_init( &ecp );
         ret = mbedtls_ecp_gen_key( grp_id, &ecp,
                                    mbedtls_ctr_drbg_random,
```
For Mbed PSA targets (e.g. K64F), we get undefined symbol error for `mbedtls_ecp_gen_key()`.

This call here is inside `#if defined(MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_KEY_PAIR)`, which is enabled by the newly added `connectivity/mbedtls/include/mbedtls/config_psa.h` if `MBEDTLS_ECP_C` is defined.

The definition of `mbedtls_ecp_gen_key()` in `connectivity/mbedtls/source/rsa.c` requires`MBEDTLS_GENPRIME`. But our script unsets it: https://github.com/ARMmbed/mbed-os/blob/47b19eb157e64c35d7164431c9b469979742b4d5/connectivity/mbedtls/tools/importer/adjust-config.sh#L85 The comment from five years ago says https://github.com/ARMmbed/mbed-os/blob/47b19eb157e64c35d7164431c9b469979742b4d5/connectivity/mbedtls/tools/importer/adjust-config.sh#L54

I wonder if we should enable `MBEDTLS_GENPRIME` globally (if current targets support it), or enable it for Mbed PSA targets, or disable `MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_KEY_PAIR` to avoid the call?

LDong-Arm replied:
It might be interesting to see whether `mbedtls_ecp_gen_key()` was called prior to the update. Though because we had something before doesn't automatically mean it was the most ideal.

Patater replied:
We'd like to avoid enabling GENPRIME_C if possible, but still be able to use PSA to import RSA keys or sign with them (at least verify would be good). I'll dig to see if this is a permitted configuration.

LDong-Arm commented on platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_MBED_PSA_SRV/mbedtls/psa_crypto_driver_wrappers.c:
```diff
@@ -0,0 +1,993 @@
+/*
```
This file needs to be added to `platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_MBED_PSA_SRV/CMakeLists.txt`.

Patater replied:
Done

wernerlewis commented on connectivity/mbedtls/include/mbedtls/config_psa.h:
```diff
@@ -0,0 +1,335 @@
+#define MBEDTLS_PSA_BUILTIN_ALG_RSA_PSS 1
+#define PSA_WANT_ALG_RSA_PSS 1
+#endif /* MBEDTLS_PKCS1_V21 */
+#define MBEDTLS_PSA_BUILTIN_KEY_TYPE_RSA_KEY_PAIR 1
```
This part of the configuration seems to be causing linker errors with `mbed-os-example-crypto`, as it is used to replace `MBEDTLS_RSA_C` and `MBEDTLS_GENPRIME` in ifdefs, but it only requires RSA_C to be enabled here. An example where it relies on `MBEDTLS_GENPRIME` (for `mbedtls_rsa_gen_key`): https://github.com/ARMmbed/mbed-os/blob/db005d04a8f81dd95c413003f0ccf6b7b3f71aff/platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_MBED_PSA_SRV/mbedtls/psa_crypto.c#L6430
Wrapping this line with an `#ifdef MBEDTLS_GENPRIME` seems to result in the same behavior of the example as before this upgrade was introduced.

Patater replied:
I came to the same conclusion and raised https://github.com/ARMmbed/mbedtls/pull/4513

We can workaround this issue until the fix is available upstream.

Patater committed:
tls: Add fix for Mbed TLS configuration issue

Until we have a fix for https://github.com/ARMmbed/mbedtls/issues/4512,
we need to patch the fix during import time. Otherwise, we run into
linker errors when PSA attempts to use RSA key generation, which we've
excluded.

This patch is extracted from
https://github.com/ARMmbed/mbedtls/pull/4513

LDong-Arm reviewed (changes requested):
@Patater Thanks, the fix looks good to me.

> @Patater We need to backport a number of definitions to TF-M's PSA headers. I encountered the same issue when trying to compile for CYTFM_064B0S2_4343W and ARM_MUSCA_S1.
> 
> To reproduce the issue, either use Mbed CLI 1, or add `mbed-mbedtls` to an application's `CMakeLists.txt` if using `mbed-tools`. Here's a quick fix: [LDong-Arm@cfe2c48](https://github.com/LDong-Arm/mbed-os/commit/cfe2c48de8b55e987dbfa5c27353fceb06d03a55) (I haven't polished the commit message yet.)
> 
> Alternatively, should we force use the Mbed TLS version of all PSA headers and remove ones from TF-M?

We still need to address this for TF-M targets. They are similar to the `mbedtls_ecc_group_to_psa()` issue we worked around previously, but now we have multiple definitions to handle instead of just one...

LDong-Arm committed:
psa: Add mbedtls_svc_key_id.h

In order for Mbed TLS to use the PSA Crypto API, definitions of
`MBEDTLS_SVC_KEY_ID_INIT`, `mbedtls_svc_key_id_t` and
`mbedtls_svc_key_id_is_null()` need to be present but are not provided
by the PSA headers from TF-M.

To solve this issue, this commit copies those definitions from Mbed
TLS's original `psa/crypto_types.h` and `psa/crypto_values.h` into a
separate `mbedtls_svc_key_id.h` for TF-M PSA.

LDong-Arm committed:
psa: Include mbedtls_svc_key_id.h for TF-M

We have added definitions that are needed by Mbed TLS's PSK key exchange
but missing from TF-M's PSA to `mbedtls_svc_key_id.h`. To pick up those
definitions, TF-M's `psa/crypto_values.h' needs to include
`mbedtls_svc_key_id.h`.

LDong-Arm committed:
psa: Add PSA_ALG_ECB_NO_PADDING to TF-M v1.0

TF-M v1.0 implements an older version of PSA and does not have the macro
`PSA_ALG_ECB_NO_PADDING` required by
`mbedtls_psa_translate_cipher_mode()` in Mbed TLS v2.25.0. Copy this
macro from Mbed TLS to fix the issue.

LDong-Arm reviewed (approved):
LGTM. Triggered https://jenkins-mbedos.oss.arm.com/blue/organizations/jenkins/mbed-os-ci_tfm-nightly/detail/mbed-os-ci_tfm-nightly/121/pipeline to test mbed-os-tf-m-regression-tests on CI boards.

LDong-Arm committed:
psa: Add missing inclusion of crypto_types.h

val_client_defs.h includes crypto_values.h, but the latter requires
type definitions from crypto_types.h.

LDong-Arm committed:
psa: Fix test detection for Mbed CLI 1

A Greentea test is detectable by Mbed CLI 1 only if it's two-levels
deep inside a `TESTS` directory, e.g. `TESTS/foo/bar/main.cpp`. But
several Mbed OS PSA tests are only one-level deep. This commit fixes
the issue by adding an extra level of directory.

LDong-Arm committed:
psa: Remove outdated macro checks in tests

The macros `TARGET_PSA` and `COMPONENT_PSA_SRV_IPC` no longer exist.
The former is replaced by `COMPONENT_PSA` which is also a directory
where the tests are located, so its check can be assumed true.
The latter is not applicable to Mbed OS PSA and can be assumed false.

Note: The entropy_inject test is skipped by default unless a user
manually configures the required `MBEDTLS_ENTROPY_NV_SEED`.

LDong-Arm reviewed (approved):
LGTM
